crates/rust-analyzer/src/global_state.rs

```diff
@@ -197,37 +197,48 @@ impl GlobalState {
             // We need to fix up the changed events a bit, if we have a create or modify for a file
             // id that is followed by a delete we actually no longer observe the file text from the
             // create or modify which may cause problems later on
+            let mut collapsed_create_delete = false;
             changed_files.dedup_by(|a, b| {
                 use vfs::ChangeKind::*;
 
+                let has_collapsed_create_delete = mem::replace(&mut collapsed_create_delete, false);
+
                 if a.file_id != b.file_id {
                     return false;
                 }
 
-                match (a.change_kind, b.change_kind) {
+                // true => delete the second element (a), we swap them here as they are inverted by dedup_by
+                match (b.change_kind, a.change_kind) {
                     // duplicate can be merged
                     (Create, Create) | (Modify, Modify) | (Delete, Delete) => true,
                     // just leave the create, modify is irrelevant
-                    (Create, Modify) => {
-                        std::mem::swap(a, b);
+                    (Create, Modify) => true,
+                    // modify becomes irrelevant if the file is deleted
+                    (Modify, Delete) => {
+                        mem::swap(a, b);
+                        true
+                    }
+                    // Remove the create message, and in the following loop, also remove the delete
```

Comment:
If the `Delete` is at the end of the list, it seems like it wouldn't get removed with this logic